README.md
# attercop

[![PyPI](https://img.shields.io/pypi/v/attercop?color=gr)](https://pypi.org/project/attercop/#description)
![PyPI - Python Version](https://img.shields.io/badge/dynamic/json?query=info.requires_python&label=python&url=https%3A%2F%2Fpypi.org%2Fpypi%2Fattercop%2Fjson)
![PyPI - License](https://img.shields.io/pypi/l/attercop)
![Code Style - Black](https://img.shields.io/badge/code%20style-black-black)

Your friendly natural language shell command generation tool!

## Overview

Attercop is a micro command line tool that generates shell commands from natural language, by way of an external large language model. Written in Python, it uses the OpenAI API to map a descriptive English language query to a command or sequence of commands, emphasizing usage of standard coreutils and tailored to the user's shell - it does its best to support bash, fish, zsh and more. For example:

```bash
$ attercop "List all the files in this directory, filtering out the ones that are not directories, and then sort them by size, largest first."
(1/1) ls -l | grep ^d | sort -k5 -n -r
```

In its default interactive mode, attercop will let you cycle through multiple options with `Tab` if the prompt is somewhat ambiguous, and either confirm and execute the output with `y` or `Enter`, copy it to the clipboard with `c`, or outright reject the command with `q`, `Ctrl+C` or `Ctrl+D`. Attercop also provides direct execution, copy and print modes to support a variety of workflows, as well as a number of optional parameters to tweak the command generation, such as the verbosity of command flags with `-v`, temperature with `-t` (essentially how creative it will get) and the max command output length `-m` as measured in GPT tokens.

## Installation

Attercop is available on PyPI, so as long as you're running Python >=3.7.1, you can install it with pip:

```bash
pip install attercop
```

You will need to set the environment variable `OPENAI_API_KEY` to your personal OpenAI API key. [You can get one here](https://beta.openai.com/) - as of January 2023, as a new member you will get plenty of free credits to play with. [For more on pricing you can see their page here.](https://openai.com/api/pricing/). Optionally you may also pass the API key as a command line argument with the `-K` or `--api-key` flag.

For developers, you can also install from source by running `pip install -e .[lint,test]` in the root directory after cloning the repository. Tests can be run locally with `pytest` and linting with `black` if you include both optional dependencies.

## Usage

Once you've installed the tool, you can just run the `attercop` command:

```bash
$ attercop "Find all files in this and subdirectories ending with the extension .txt"
(1/1): find . -name "*.txt"
```

It will hit the OpenAI API and come back with one of several prompts, depending on the max_prompts parameter.

### Interactive Mode

By default, attercop will run in interactive mode, in which the user will be prompted for confirmation before any command is executed or copies. The interactive mode controls are as follows:

| Command           | Key                         |
|-------------------|-----------------------------|
| Cycle Command     | `Tab`                       |
| Execute           | `y` \| `Enter`              |
| Copy to Clipboard | `c`                         |
| Reject            | `q` \| `Ctrl+C` \| `Ctrl+D` |

### Execute Mode

Optionally, you can also include the `-X` or `--execute` flag to execute the command **immediately without confirmation**. Please approach with great caution! Although direct execution mode may be useful for certain workflows, it should nonetheless be approached carefully as you are asking attercop to execute arbitrary commands on your system **without any confirmation!** attercop does its best to identify dangerous (removing files, curling inputs for execution, etc.) or privileged (sudo, su...) commands and will exit direct execution mode if identified as such, but it is important to note that this is not a guarantee of safety and it should nonetheless be used with great caution and care. If you are unsure, it is always recommended to use the default interactive mode.

### Copy Mode

Direct copy mode with the `-c` or `--copy` flag similarly copies the output directly to your clipboard without confirmation. This is more lenient and will not exit if a dangerous command is identified, though it will print a warning to the console (via stderr) - please use caution and review the command before executing!

### Print Mode

Print mode with the `-p` or `--print` flag outputs the generated command directly to stdout without prompting for user input. Similarly to copy mode, will not exit if a dangerous command is executed, but will print a warning to stderr.

## Help and Options

For the full menu of options and optional arguments, including various parameters to tweak the API call, see `attercop --help`:

```
$ attercop --help
usage: attercop [-h] [-X | -c | -p] [-v] [-n [1, 10]] [-t [0, 2]] [-m [1, 1024]] [-M MODEL] [-K API_KEY] [-s SHELL] prompt

Generate a command or chain of shell commands from a natural language prompt. In the default interactive mode, you can cycle through commands with tab, execute with enter or y, copy to the clipboard with c, or quit with q.

positional arguments:
  prompt                The English-language prompt to use for the GPT completion.

options:
  -h, --help            show this help message and exit
  -X, --execute         Execute mode: runs the generated command immediately without confirmation! Tries to identify dangerous or privileged commands and exit, but should nonetheless be used with great caution.
  -c, --copy            Copy mode: copies the generated command to the clipboard immediately without prompting for user input.
  -p, --print           Print mode: outputs the generated command directly to stdout without prompting for user input.
  -v, --verbose         Explicitly prefer verbose flags for the generated command, ie. `--help` instead of `-h`.
  -n [1, 10], --num-prompts [1, 10]
                        The maximum number of alternative prompts to generate. Defaults to 3.
  -t [0, 2], --temperature [0, 2]
                        Higher values will result in more diverse completions, but lower values will generally result in more sensible ones. Defaults to 0.
  -m [1, 1024], --max-tokens [1, 1024]
                        The maximum number of tokens to generate per prompt. Defaults to 100.
  -M MODEL, --model MODEL
                        The GPT model to use. Defaults to text-davinci-003, the latest and most powerful model generally available.
  -K API_KEY, --api-key API_KEY
                        Manually specify an OpenAI API key to use. If none provided, will default to the OPENAI_API_KEY environment variable.
  -s SHELL, --shell SHELL
                        The shell to use for the generated command, ie. bash, zsh, fish, etc. If none provided, will look for a SHELL environment variable if available, or otherwise default to bash.
```

## FAQ

### What is a token and how do I know how many I need?

Tokens are the basic unit of text in the GPT-3 API. They represent a sort of piece of a word or, in our case, a kind of "stem" of a command. [You can read more about the concept here](https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them) - the default max tokens is 100, which is generally enough to generate all but the most complex of chained commands. If you are doing something super crazy, you may need to increase this value - you can use the `-m` or `--max-tokens` flag to set this value manually.

### What about support for other LLM providers?

I would definitely prefer this tool to be agnostic to the LLM provider. Right now OpenAI is pretty much the only game in town but if you have an idea to add support for another provider (including perhaps a locally-hosted GPT variant) please open an issue or a pull request! Also please do open an issue if you have any other suggestions, identify any bugs or have any feedback more broadly.

### Why GPLv3 and not MIT?

A strange bearded man came to me in a dream and told me to do so before launching into a diatribe about the immorality of buying parrots. Not sure what that last part was all about but hey, I'm not one to tempt fate.

### Why is this tool called attercop?

Mainly I just thought it sounded cool! If you really must know, its from an old English word for spider, and [referenced in a song Bilbo sings in The Hobbit to lure away some spiders from his friends who are stuck in a tree.](https://tolkiengateway.net/wiki/Old_fat_spider_spinning_in_a_tree!) So the connection with the tool is quite tenuous but I suppose you could draw some kind of analogy to the web of manpages that one must traverse to find the right flags for a command, ha ha.


CHANGES.md
# Changes

## v0.3.4

* Add `-delete` to the list of dangerous flags, eg. `find . -delete` and similar. Add test as well
* Tweak the dangerous and privileged flag display in interactive mode and use red text with colorama
* Update Python version requirement in README

## v0.3.3

* Improve verbosity behaviour by further specifying prompt and including an additional example

## v0.3.2

* Replace match statement with if/else to support older Python versions back to 3.7.1 (bounded by OpenAI python library)
* Remove newline characters being appended to the end of the command in print mode
* Add tests for print mode branch
* Add changes tracking file for tagged releases
* Add workflow to publish tagged releases to PyPI

## v0.3.1

* Add new print mode to output the command directly to stdout, complete with warnings to stderr for flagged commands
* Improve the README as well as the help prompt

## v0.3.0

* Add direct execution and copy mode to allow for usage in scripts etc. Due to the danger involved here, I added some somewhat stern warnings in the README and help prompts, as well as...
* New `dangerous` and `privileged` flags to try to cover most of the potential pain zones - think sudo rm -rf / and various permutations of deletions and privilege escalations. Will show to the user and refuses to run direct execution mode
* A test suite! Covers most control flow branches, not totally comprehensive yet but I'm verifying execution and copy behaviour in direct and interactive mode, as well as flag parsing as described above
* Github workflow to insist on black linting as well as the passage of the above tests
* Overall misc. improvements to the main execution loop, arguments etc.

## v0.2.0

* Add a verbosity flag `-v | --verbose`, credit to [user chris37879 on reddit for the idea](https://www.reddit.com/r/programming/comments/106q5jh/comment/j3j63nw/?utm_source=reddit&utm_medium=web2x&context=3)! This will encourage the completion to include full verbose flags when available, ie `--help` instead of `-h`.
* Add a shell flag `-s | --shell` and automatic shell detection, to hopefully account for some of the nuance between how ie. bash and zsh or fish handle environment variables, etc.
* Also add an optional API key flag `-K | --api-key` in case people don't want to use the environment variable. Will still default to checking for that env var, and will now throw an exception if nothing is found.
* Fix bug to ensure arguments are actually being used in the API call!
* Fix a potential issue where if duplicate commands are returned in a staggered order, the command cycling may select duplicates instead of the unique values

## v0.1.2

* Documentation and metadata tweaks

## v0.1.1

* Upgrades the default model to `davinci-003` for better performance
* Adds a new copy to clipboard command
* Enhances the README and help text with more details on commands
* Very slight tweak to prompt

## v0.1.0

* Initial release!


tests/test_args.py
import pytest

from attercop import attercop


def test_standard_args():
    args = attercop.parse_args(["My name jeff", "-c"])
    assert args.prompt == "My name jeff"
    assert args.copy == True
    assert args.execute == False
    # Defaults
    assert args.model == "text-davinci-003"
    assert args.temperature == 0.0
    assert args.num_prompts == 1  # Copy mode only generates one prompt

    args = attercop.parse_args(
        [
            "Delete the computer",
            "-X",
            "--model",
            "text-davinci-2020-05-03",
            "--temperature",
            "0.5",
        ]
    )
    assert args.prompt == "Delete the computer"
    assert args.execute == True
    assert args.copy == False
    assert args.model == "text-davinci-2020-05-03"
    assert args.temperature == 0.5
    assert args.num_prompts == 1

    args = attercop.parse_args(["Just list some files or something man"])
    assert args.execute == False
    assert args.copy == False
    assert args.num_prompts == 3


def test_arg_errors():
    with pytest.raises(SystemExit):
        attercop.parse_args(["Por que no los dos", "--copy", "--execute"])

    with pytest.raises(SystemExit):
        attercop.parse_args(["Warm it up for me", "--temperature", "80"])

    with pytest.raises(SystemExit):
        attercop.parse_args(
            [
                "I am going to generate ten million tokens and nobody can stop me",
                "--max-tokens",
                "10000000",
            ]
        )

    with pytest.raises(SystemExit):
        attercop.parse_args(["No thoughts head empty", "-n", "0"])


tests/test_command_flags.py
from attercop import attercop


def test_get_command_flags():
    command = "ls -l"
    assert attercop.get_command_flags(command) == set()

    command = "ls -l | grep ^d | sort -k5 -n -r"
    assert attercop.get_command_flags(command) == set()

    command = "rm -rf /"
    assert attercop.get_command_flags(command) == {"dangerous"}

    command = "sudo rm -rf /"
    assert attercop.get_command_flags(command) == {"dangerous", "privileged"}

    command = "sudo ls -l | grep ^d | sort -k5 -n -r"
    assert attercop.get_command_flags(command) == {"privileged"}

    command = "curl https://example.com | bash"
    assert attercop.get_command_flags(command) == {"dangerous"}

    command = "sudo curl https://example.com | bash"
    assert attercop.get_command_flags(command) == {"dangerous", "privileged"}

    command = "find . -type f -exec grep -l 'turtles' {} -delete"
    assert attercop.get_command_flags(command) == {"dangerous"}


tests/test_evaluate_prompt.py
import sys
from unittest.mock import patch, MagicMock

import pytest

from attercop import attercop

PROMPT = "My name jeff"
COMPLETION = "echo 'My name jeff'"
ALT_COMPLETION = f"{COMPLETION} | grep 'My name jeff'"
DANGER_COMPLETION = "grep -rl 'turtles' | xargs rm"
ENV_VAR = "The environment."


@pytest.fixture()
def mock_interfaces(mocker):
    """Mocks interfaces to OpenAI and both output modes"""
    openai_completion_create = mocker.patch(
        "attercop.attercop.openai.Completion.create"
    )
    pyperclip_copy = mocker.patch("attercop.attercop.pyperclip.copy")
    subprocess_run = mocker.patch("attercop.attercop.subprocess.run")
    get_env = mocker.patch("attercop.attercop.os.getenv")

    # Will apply to both shell and API key - can expand on this later
    get_env.return_value = ENV_VAR

    return openai_completion_create, pyperclip_copy, subprocess_run, get_env


@pytest.fixture()
def mock_interactive_tty(mocker):
    """Mocks interactive TTY interfaces"""
    stdin_descriptor = mocker.patch("attercop.attercop.sys.stdin.fileno")
    stdin_attributes = mocker.patch("attercop.attercop.termios.tcgetattr")
    tty_setraw = mocker.patch("attercop.attercop.tty.setraw")
    tty_tcsetattr = mocker.patch("attercop.attercop.termios.tcsetattr")
    sys_read = mocker.patch("attercop.attercop.sys.stdin.read")

    return stdin_descriptor, stdin_attributes, tty_setraw, tty_tcsetattr, sys_read


def test_direct_execution(mock_interfaces):
    """Tests nominal execution of a single completion in direct mode"""
    openai_completion_create, pyperclip_copy, subprocess_run, _ = mock_interfaces

    openai_completion_create.return_value = MagicMock(
        choices=[MagicMock(text=COMPLETION)]
    )

    sys.argv = [sys.argv[0], PROMPT, "-X"]
    attercop.evaluate_prompt()

    openai_completion_create.assert_called_once()
    subprocess_run.assert_called_once_with(COMPLETION, shell=True)
    pyperclip_copy.assert_not_called()


def test_dangerous_direct_execution(mock_interfaces):
    """Test abort of dangerous execution in direct mode"""
    openai_completion_create, pyperclip_copy, subprocess_run, _ = mock_interfaces

    openai_completion_create.return_value = MagicMock(
        choices=[MagicMock(text=DANGER_COMPLETION)]
    )

    sys.argv = [sys.argv[0], PROMPT, "-X"]

    with pytest.raises(SystemExit) as pytest_wrapped_e:
        attercop.evaluate_prompt()

    openai_completion_create.assert_called_once()
    subprocess_run.assert_not_called()
    pyperclip_copy.assert_not_called()


def test_direct_copy(mock_interfaces):
    """Tests nominal copy of a single completion in direct mode"""
    openai_completion_create, pyperclip_copy, subprocess_run, _ = mock_interfaces

    openai_completion_create.return_value = MagicMock(
        choices=[MagicMock(text=COMPLETION)]
    )

    sys.argv = [sys.argv[0], PROMPT, "-c"]
    attercop.evaluate_prompt()

    openai_completion_create.assert_called_once()
    subprocess_run.assert_not_called()
    pyperclip_copy.assert_called_once_with(COMPLETION)


def test_direct_print(mock_interfaces, capsys):
    """Tests nominal print of a single completion in direct mode"""
    openai_completion_create, pyperclip_copy, subprocess_run, _ = mock_interfaces

    openai_completion_create.return_value = MagicMock(
        choices=[MagicMock(text=COMPLETION)]
    )

    sys.argv = [sys.argv[0], PROMPT, "--print"]
    attercop.evaluate_prompt()

    output = capsys.readouterr()
    assert output.out == COMPLETION
    assert not output.err

    openai_completion_create.assert_called_once()
    subprocess_run.assert_not_called()
    pyperclip_copy.assert_not_called()


def test_dangerous_print(mock_interfaces, capsys):
    """Ensure that warning flags are raised to stderr when printing dangerous completions"""
    openai_completion_create, pyperclip_copy, subprocess_run, _ = mock_interfaces

    openai_completion_create.return_value = MagicMock(
        choices=[MagicMock(text=DANGER_COMPLETION)]
    )

    sys.argv = [sys.argv[0], PROMPT, "--print"]
    attercop.evaluate_prompt()

    output = capsys.readouterr()
    assert output.out == DANGER_COMPLETION
    assert "<dangerous>" in output.err

    openai_completion_create.assert_called_once()
    subprocess_run.assert_not_called()
    pyperclip_copy.assert_not_called()


def test_interactive_execution(mock_interfaces, mock_interactive_tty):
    """Tests nominal execution of a completion in interactive mode

    This test mocks the TTY interfaces and simulates a user cycling around
    before accepting an alternative completion for execution.
    """
    openai_completion_create, pyperclip_copy, subprocess_run, _ = mock_interfaces
    (
        stdin_descriptor,
        stdin_attributes,
        tty_setraw,
        tty_tcsetattr,
        sys_read,
    ) = mock_interactive_tty

    openai_completion_create.return_value = MagicMock(
        choices=[MagicMock(text=COMPLETION), MagicMock(text=ALT_COMPLETION)]
    )

    sys.argv = [sys.argv[0], PROMPT]
    sys_read.side_effect = ["\t"] * 3 + ["y"]

    attercop.evaluate_prompt()

    openai_completion_create.assert_called_once()
    subprocess_run.assert_called_once_with(ALT_COMPLETION, shell=True)
    pyperclip_copy.assert_not_called()

    stdin_descriptor.assert_called_once()
    stdin_attributes.assert_called_once()
    tty_setraw.assert_called_once()
    tty_tcsetattr.assert_called_once()

    assert sys_read.call_count == 4


def test_interactive_reject(mock_interfaces, mock_interactive_tty):
    """Tests nominal rejection of a completion in interactive mode"""
    openai_completion_create, pyperclip_copy, subprocess_run, _ = mock_interfaces
    (
        stdin_descriptor,
        stdin_attributes,
        tty_setraw,
        tty_tcsetattr,
        sys_read,
    ) = mock_interactive_tty

    openai_completion_create.return_value = MagicMock(
        choices=[MagicMock(text=COMPLETION)]
    )

    sys.argv = [sys.argv[0], PROMPT]
    sys_read.side_effect = ["\t", "q"]

    attercop.evaluate_prompt()

    openai_completion_create.assert_called_once()
    subprocess_run.assert_not_called()
    pyperclip_copy.assert_not_called()

    stdin_descriptor.assert_called_once()
    stdin_attributes.assert_called_once()
    tty_setraw.assert_called_once()
    tty_tcsetattr.assert_called_once()

    assert sys_read.call_count == 2


src/attercop/attercop.py
import os
import argparse
import sys
import tty
import termios
import subprocess
import pathlib

import openai
import pyperclip

EXECUTE, COPY, PRINT = "execute", "copy", "print"
CAUTION_FLAGS = {
    "dangerous": (
        "rm",
        "mv",
        "kill",
        "shutdown",
        "reboot",
        "poweroff",
        "halt",
        "dd",
        "wget",
        "curl",
    ),
    "privileged": ("sudo", "su", "doas", "pkexec", "gksudo", "gksu", "kdesudo", "ksu"),
}


def parse_args(args_override: list = None) -> argparse.Namespace:
    """Parse command line arguments.

    Will exit if the arguments are invalid.
    """
    parser = argparse.ArgumentParser(
        description="Generate a command or chain of shell commands from a natural language prompt. In the default interactive mode, you can cycle through commands with tab, execute with enter or y, copy to the clipboard with c, or quit with q."
    )
    parser.add_argument(
        "prompt",
        type=str,
        help="The English-language prompt to use for the GPT completion.",
    )
    execution_mode = (
        parser.add_mutually_exclusive_group()
    )  # Interactive by default, otherwise only one of these can be set
    execution_mode.add_argument(
        "-X",
        f"--{EXECUTE}",
        action="store_true",
        help="Execute mode: runs the generated command immediately without confirmation! Tries to identify dangerous or privileged commands and exit, but should nonetheless be used with great caution.",
    )
    execution_mode.add_argument(
        "-c",
        f"--{COPY}",
        action="store_true",
        help="Copy mode: copies the generated command to the clipboard immediately without prompting for user input.",
    )
    execution_mode.add_argument(
        "-p",
        f"--{PRINT}",
        action="store_true",
        help="Print mode: outputs the generated command directly to stdout without prompting for user input.",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Explicitly prefer verbose flags for the generated command, ie. `--help` instead of `-h`.",
    )
    parser.add_argument(
        "-n",
        "--num-prompts",
        type=int,
        default=3,
        metavar=[1, 10],
        help="The maximum number of alternative prompts to generate. Defaults to 3.",
    )
    parser.add_argument(
        "-t",
        "--temperature",
        type=float,
        default=0,
        metavar=[0, 2],
        help="Higher values will result in more diverse completions, but lower values will generally result in more sensible ones. Defaults to 0.",
    )
    parser.add_argument(
        "-m",
        "--max-tokens",
        type=int,
        default=100,
        metavar=[1, 1024],
        help="The maximum number of tokens to generate per prompt. Defaults to 100.",
    )
    parser.add_argument(
        "-M",
        "--model",
        type=str,
        default="text-davinci-003",
        help="The GPT model to use. Defaults to text-davinci-003, the latest and most powerful model generally available.",
    )
    parser.add_argument(
        "-K",
        "--api-key",
        type=str,
        default=None,
        help="Manually specify an OpenAI API key to use. If none provided, will default to the OPENAI_API_KEY environment variable.",
    )
    parser.add_argument(
        "-s",
        "--shell",
        type=str,
        default=None,
        help="The shell to use for the generated command, ie. bash, zsh, fish, etc. If none provided, will look for a SHELL environment variable if available, or otherwise default to bash.",
    )
    args = parser.parse_args(args_override)

    if args.temperature < 0 or args.temperature > 2:
        parser.error("Temperature must be between 0.0 and 2.0")
    elif args.num_prompts < 1 or args.num_prompts > 10:
        parser.error("Number of prompts must be between 1 and 10")
    elif args.max_tokens < 1 or args.max_tokens > 1024:
        parser.error("Maximum tokens must be between 1 and 1024")

    if args.execute or args.copy or args.print:
        args.num_prompts = 1  # Save on tokens in any non-interactive mode

    return args


def generate_prompt(args: argparse.Namespace) -> str:
    """Generate the GPT completion prompt from the provided command line arguments

    Determines the user's shell according to the precedence described in the shell argument help text
    If verbose is set, will also prefer verbose flags for the generated command, ie. `--help` instead of `-h`
    """

    shell = None
    if args.shell:
        shell = args.shell
    else:
        env_shell = os.getenv("SHELL")
        if env_shell:
            shell = pathlib.Path(env_shell).name  # ie. /bin/bash -> bash
        else:
            shell = "bash"

    verbosity_clause = (
        " Use verbose flags absolutely wherever possible, ie. --help instead of -h, head --lines=3 instead of n=3, etc.\n"
        if args.verbose
        else "\n"
    )
    sort_clause = (  # Include an additional example to get through GPT's thick skull
        "sort --key=5 --numeric-sort --reverse\nPrompt: Create an annotated git tag for version v1.0\nCommand: git tag --annotate v1.0 --message 'Version 1.0'\n"
        if args.verbose
        else "sort -k5 -n -r\n"
    )
    prompt = (
        f"Convert a prompt into a working programmatic {shell} shell command or chain of commands, making use of standard GNU tools and common Unix idioms."
        + verbosity_clause
        + "Prompt: List all the files in this directory, filtering out the ones that are not directories, and then sort them by size, largest first.\n"
        + "Command: ls -l | grep ^d | "
        + sort_clause
        + f"Prompt: {args.prompt}\n"
        + "Command: "
    )

    return prompt


def get_command_flags(command: str) -> set:
    """Identify dangerous or privileged commands

    Put together this puts reasonable oversight on common commands that could
    potentially delete files, format disks, or run untrusted code
    """

    flags = set()

    for concern in CAUTION_FLAGS:
        for flag in CAUTION_FLAGS[concern]:
            if f"{flag} " in command or command.endswith(flag):
                flags.add(concern)

    return flags


def evaluate_prompt() -> None:
    """Evaluate the prompt and generate a command or chain of commands

    This is the main function that handles the command line arguments, the GPT completion, and the
    interactive command selection loop.
    """
    try:
        args = parse_args()

        if args.api_key:
            openai.api_key = args.api_key
        else:
            openai.api_key = os.getenv("OPENAI_API_KEY")
        if not openai.api_key:
            raise ValueError(
                "No OpenAI API key provided. Please provide an API key either via the -K/--api-key flag, or by setting OPENAI_API_KEY environment variable in your shell. "
                + "See https://openai.com/docs/developer-quickstart for more information."
            )

        prompt = generate_prompt(args)

        response = openai.Completion.create(
            model=args.model,
            prompt=prompt,
            temperature=args.temperature,
            max_tokens=args.max_tokens,
            top_p=1.0,
            n=args.num_prompts,
            frequency_penalty=0.2,
            presence_penalty=0.0,
            stop=["\n"],
        )

        # Deduplicate choices while still preserving ordering (dict is insertion ordered)
        outputs = list(
            {
                choice.text.strip(): get_command_flags(choice.text)
                for choice in response.choices
            }.items()
        )

        if not "".join([output[0] for output in outputs]):
            raise ValueError(
                "No outputs generated. Please try again with a different prompt or model."
            )

        # Queue up the first command for execution or copying
        selected_query = 0
        output, flags = outputs[selected_query]

        # May skip interactive loop in some conditions. Otherwise, enter interactive loop
        if args.execute:
            if flags:
                raise ValueError(
                    f"Command `{output}` triggered cautionary flags <{', '.join(flags)}>\nPlease run in interactive mode to review the command for manual execution."
                )
            action = EXECUTE
        elif args.copy or args.print:
            if flags:
                print(
                    f"Command `{output}` triggered cautionary flags <{', '.join(flags)}>\nPlease review before running!",
                    file=sys.stderr,
                )
            action = COPY if args.copy else PRINT
        else:
            action = None

    except (ValueError, openai.error.OpenAIError) as error:
        sys.exit(f"Error: {error}")

    # Interactive command selection loop
    if not action:
        stdin_descriptor = sys.stdin.fileno()
        stdin_attributes = termios.tcgetattr(stdin_descriptor)
        tty.setraw(stdin_descriptor)

    try:
        while not action:
            print(
                f"({selected_query + 1}/{len(outputs)}{' ' + str(flags) if flags else ''}): {output}",
                end="\r",
            )
            ch = sys.stdin.read(1)

            if ch == "y" or ch == "\r":
                action = EXECUTE
            elif ch == "c":
                action = COPY
            elif ch == "\t":
                selected_query = (selected_query + 1) % len(outputs)
                output, flags = outputs[selected_query]
            elif (
                ch == "q" or ch == "\x03" or ch == "\x04"
            ):  # SIGKILL & SIGTERM - not ideal tty handling but hey it works
                break

            if not action:
                print("\033[K", end="")  # Clear line
    finally:
        if not args.execute and not args.copy and not args.print:
            termios.tcsetattr(stdin_descriptor, termios.TCSADRAIN, stdin_attributes)

    if action == EXECUTE:
        print(f"Executing: {output}")
        subprocess.run(output, shell=True)
    elif action == COPY:
        print(f"Copying to clipboard: {output}")
        pyperclip.copy(output)
    elif action == PRINT:
        print(output, end="")


if __name__ == "__main__":
    evaluate_prompt()
